docs/modules/ROOT/pages/references/style-guide.adoc

```diff
@@ -0,0 +1,273 @@
++
+[source,jsonnet]
+----
+{
+  // CORRECT
+  local myVariable = 3,
+  result: myVariable + 1,
+
+  // INCORRECT
+  myVariable:: 3,
+  result: $.myVariable + 1,
+}
+----
+
+=== Line Length
+
+* Limit lines to 120 characters.
+* The only exceptions are import statements and URLs (although even for those, try to keep them under 120 chars).
+
+=== Spacing and Indentation
+
+* Put one space before and after operators.
++
+[source,jsonnet]
+----
+local c = a + b;
+----
+
+* Put one space after commas.
++
+[source,jsonnet]
+----
+["a", "b", "c"] // CORRECT
+
+["a","b","c"] // INCORRECT
+----
+
+* Put one space after colons.
++
+[source,jsonnet]
+----
+{
+  // CORRECT
+  foo:: "bar",
+  baz: "taz",
+  { hello: "world" },
+
+  // INCORRECT
+  foo :: "bar",
+  baz:"taz",
+  { hello : "world" },
+}
+----
+
+* Put one space or line brake after opening curly brackets and before closing curly brackets.
++
+[source,jsonnet]
+----
+// CORRECT
+local foo = { hello: "world" };
+local bar = {
+  hello: "world",
+};
+
+// INCORRECT
+local baz = {hello: "world"};
+----
+
+* Use 2-space indentation in general.
+
+* Only method or class parameter declarations use 4-space indentation, to visually differentiate parameters from method body.
++
+[source,jsonnet]
+----
+// CORRECT
+local multiply(
+    number1,
+    number2) = {
+  result: number1 * number 2
+}
+----
+
+* Omit vertical alignment.
+  They draw attention to the wrong parts of the code and result in hard to review pull requests due to the white space changes.
++
+[source,jsonnet]
+----
+// Don't align vertically
+local plus     = "+";
+local minus    = "-";
+local multiply = "*";
+
+// Do the following
+local plus = "+";
+local minus = "-";
+local multiply = "*";
+----
+
+=== Blank Lines (Vertical Whitespace)
+
+* A single blank line appears:
+** Within method bodies, as needed to create logical groupings of statements.
+** Optionally before the first member or after the last member of a class or method.
+* Use one or two blank line(s) to separate class definitions.
+* Excessive number of blank lines is discouraged.
+
+=== Defining and Using Abstractions
+
+==== Defining Classes
+
+* Rather than defining a concrete JSON file, it's often useful to define a template which takes some set of parameters before being materialized into JSON.
+  We can liken named functions which take a set of parameters and result in a fixed scheme to ``classes'' in object-oriented languages, and so we will use that terminology.
+* When defining a class, use the following syntax:
++
+[source,jsonnet]
+----
+local newAnimal(name, age) = {
+  name: name,
+  age: age,
+};
+{
+  newAnimal:: newAnimal,
+}
+----
+
+* Returning a dictionary with a `newXxx` method (rather than just returning the constructor directly) allows exposing constants, static methods, or related class constructors from the same file.
+  In other words, it allows extending this class in the future without refactoring all downstream consumers.
```

Comment:
```suggestion
* When writing libraries, always return a single object encapsulating any methods instead of returning a single method.
  This allows returning multiple values (constants, static methods or class constructors) from a single library.
  Additionally this ensures libraries remain extensible without having to refactor all consumers.
```